core/icons/mutator_icon.ts

```diff
@@ -0,0 +1,358 @@
+      'renderer': this.sourceBlock.workspace.options.renderer,
+      'rendererOverrides':
+        this.sourceBlock.workspace.options.rendererOverrides ?? undefined,
+    };
+
+    if (this.flyoutBlockTypes.length) {
+      options.toolbox = {
+        'kind': 'flyoutToolbox',
+        'contents': this.flyoutBlockTypes.map((type) => ({
+          'kind': 'block',
+          'type': type,
+        })),
+      };
+    }
+
+    return options;
+  }
+
+  /**
+   * @returns the location the bubble should be anchored to.
+   *     I.E. the middle of this icon.
+   */
+  private getAnchorLocation(): Coordinate {
+    const midIcon = this.SIZE / 2;
+    return Coordinate.sum(
+      this.workspaceLocation,
+      new Coordinate(midIcon, midIcon)
+    );
+  }
+
+  /**
+   * @returns the rect the bubble should avoid overlapping.
+   *     I.E. the block that owns this icon.
+   */
+  private getBubbleOwnerRect(): Rect {
+    const bbox = this.sourceBlock.getSvgRoot().getBBox();
+    return new Rect(bbox.y, bbox.y + bbox.height, bbox.x, bbox.x + bbox.width);
+  }
+
+  /** Decomposes the source block to create blocks in the mini workspace. */
+  private createRootBlock() {
+    this.rootBlock = this.sourceBlock.decompose!(
+      this.miniWorkspaceBubble!.getWorkspace()
+    )!;
+
+    for (const child of this.rootBlock.getDescendants(false)) {
+      child.queueRender();
+    }
+
+    this.rootBlock.setMovable(false);
+    this.rootBlock.setDeletable(false);
+
+    const flyoutWidth =
+      this.miniWorkspaceBubble?.getWorkspace()?.getFlyout()?.getWidth() ?? 0;
+    this.rootBlock.moveBy(
+      this.rootBlock.RTL ? -(flyoutWidth + this.MARGIN) : this.MARGIN,
+      this.MARGIN
+    );
+  }
+
+  /** Adds a listen to the source block that triggers saving connections. */
+  private addSaveConnectionsListener() {
+    if (!this.sourceBlock.saveConnections || !this.rootBlock) return;
+    const saveConnectionsListener = () => {
+      if (!this.sourceBlock.saveConnections || !this.rootBlock) return;
+      this.sourceBlock.saveConnections(this.rootBlock);
+    };
+    saveConnectionsListener();
+    this.sourceBlock.workspace.addChangeListener(saveConnectionsListener);
+  }
+
+  /**
+   * Creates a change listener to add to the mini workspace which recomposes
+   * the block.
+   */
+  private createMiniWorkspaceChangeListener() {
+    return (e: Abstract) => {
+      if (!this.shouldIgnoreMutatorEvent(e) && !this.updateWorkspacePid) {
+        this.updateWorkspacePid = setTimeout(() => {
+          this.updateWorkspacePid = null;
+          this.recomposeSourceBlock();
+        }, 0);
+      }
+    };
+  }
+
+  /**
+   * Returns true if the given event is not one the mutator needs to
+   * care about.
+   */
+  private shouldIgnoreMutatorEvent(e: Abstract) {
+    return (
+      e.isUiEvent ||
+      e.type === eventUtils.CREATE ||
+      (e.type === eventUtils.CHANGE &&
+        (e as BlockChange).element === 'disabled')
+    );
+  }
```

Comment:
This doesn't seem to ned to be a method on `MutatorIcon`—certainly it could be a `static` method at least, or just a freestanding function.

Also, would `isIgnorableMutatorEvent` be a better name?

Reply:
Similar concerns about cluttering up our exports with static methods, but again no particular preference.

wrt `isIgnorableMutatorEvent` I tend to read things like: `if (this.shouldIgnoreMutatorEvent(e)) { doSomething(); }` as "if this (icon) should ignore (that) mutator event then do something" so "if this (event) is (an) ignorable mutator event" feels a bit strange to me.

But I'll go ahead and change it since `is`-prefixed boolean methods are more common!